spring-cloud-gateway-server-mvc/src/main/java/org/springframework/cloud/gateway/server/mvc/handler/ProxyExchangeHandlerFunction.java

```diff
@@ -84,14 +84,12 @@ private void init() {
 	@Override
 	public ServerResponse handle(ServerRequest serverRequest) {
 		URI uri = uriResolver.apply(serverRequest);
-		boolean encoded = containsEncodedQuery(serverRequest.uri(), serverRequest.params());
 		// @formatter:off
 		URI url = UriComponentsBuilder.fromUri(serverRequest.uri())
 				.scheme(uri.getScheme())
 				.host(uri.getHost())
 				.port(uri.getPort())
-				.replaceQueryParams(serverRequest.params())
```

Comment:
if you add some query params in pre filters they will not be added to request if you remove this line

Reply:
Good catch, I assumed query params were already included in `serverRequest.uri()`. I think this needs an additional integration test.

Reply:
I've been trying to fix this issue in our project aswell, and what I find to be working best is this
```java
URI url = UriComponentsBuilder.fromUri(serverRequest.uri())
            .scheme(uri.getScheme())
            .host(uri.getHost())
            .port(uri.getPort())
            .replaceQueryParams(serverRequest.params())
            .encode()
            .build()
            .toUri();
```
It solves all cases we have found so far.